mobzy-systems/src/main/kotlin/com/mineinabyss/mobzy/systems/listeners/LeashingListener.kt

```diff
@@ -0,0 +1,87 @@
+import org.bukkit.entity.LivingEntity
+import org.bukkit.entity.PufferFish
+import org.bukkit.event.EventHandler
+import org.bukkit.event.Listener
+import org.bukkit.event.entity.EntityDamageEvent
+import org.bukkit.event.entity.PlayerLeashEntityEvent
+import org.bukkit.event.player.PlayerUnleashEntityEvent
+import org.bukkit.potion.PotionEffectType
+
+object LeashingListener : Listener {
+
+    // Cancel Pufferfish change for leash-entities
+    @EventHandler
+    fun PufferFishStateChangeEvent.onLeash() {
+        isCancelled = entity.toGeary().has<LeashEntity>()
+    }
+
+    /** Handling leashing of entities with [ModelEngineComponent] */
+    @EventHandler
+    fun PlayerLeashEntityEvent.onLeashingMob() {
+        val gearyEntity = entity.toGearyOrNull() ?: return
+        val modelEntity = entity.toModelEntity() ?: return
+
+        // Leash is tied to the ModelEngine BaseEntity.
+        // This makes it exist for interaction but also invisible
+        // Spawn pufferfish, mount it to entity-bone and set leash-holder
+        gearyEntity.with { componentEntity: ModelEngineComponent ->
+            if (!componentEntity.leashable) return
+            if ((entity as? LivingEntity ?: return).isLeashed) return
+
+            val mount = modelEntity.mountHandler
+            val pufferFish = player.world.spawnEntity(entity.location, EntityType.PUFFERFISH) as PufferFish
+            pufferFish.puffState = 0
+            pufferFish.setAI(false)
+            pufferFish.isInvisible = true
+            pufferFish.isInvulnerable = true
+            pufferFish.toGeary().setPersisting(LeashEntity())
+
+            mount.setCanCarryPassenger(true)
+            mount.addPassenger("leash", pufferFish)
+            (entity as LivingEntity).setLeashHolder(player)
+            (pufferFish as LivingEntity).setLeashHolder(player)
+        }
+    }
+
+    /** Handle unleashing of entities with [ModelEngineComponent.leashable] */
+    @EventHandler
+    fun PlayerUnleashEntityEvent.onUnleashMob() {
+        val gearyEntity = entity.toGearyOrNull() ?: return
+        val modelEntity = entity.toModelEntity() ?: return
+
+        // Leash is tied to the ModelEngine BaseEntity and rendered via a mounted pufferfish.
+        // Since we unleash, we unleash BaseEntity and remove pufferfish
+        gearyEntity.with { componentEntity: ModelEngineComponent ->
+            if (!componentEntity.leashable) return
+            modelEntity.isInvisible = true
+            modelEntity.entity.removePotionEffect(PotionEffectType.INVISIBILITY)
+            modelEntity.mountHandler.getPassengersOnBone("leash")
+                .filter { it.toGeary().has<LeashEntity>() }.forEach {
+                    (it as LivingEntity).setLeashHolder(null)
+                    it.remove()
+                }
+        }
+    }
+
+    /** Handle unleashing of entities with [ModelEngineComponent.leashable] on death */
+    @EventHandler
+    fun EntityDamageEvent.unleashOnDeath() {
```

Comment:
Should we have another event/replace this event for entity removed from world? It sounds like a far more consistent way to remove the pufferfish

Reply:
It does need to be ran before the entity is removed though. Hence why EntityDeath doesnt work.
This might also be why the Bossbar issue for mobs are happening. As it is ran on mobs marked for removal
Not sure if this means they HAVE been removed when that runs but

Reply:
Tried now with EntityRemoveFromWorldEvent, but it fires for each bone on the ME-entity. Thus none of them are "modelengine mobs" and it never works as it should. I think this might be the issue aswell with EntityDeathEvent? not sure

Reply:
Very odd that event fires for each bone... are you saying it never fires on anything considered an ME entity? Id assume you can just filter out the bones and use what u currently have? All sounds weird

Reply:
The problem is it doesnt treat it as one entity and on their own they never count as a modelengine entity or something
Tried filtering and checking if a bone was said thing but no all returned false.
I could retry it but it seems to just be the case when entities dont actually exist

Hence why mounted mobs get fucked on relog/server shutdown and bossbar gets borked on entity death? idk

Reply:
Hmm I suppose then modelengine unregisters it before this event fires. You could look around their code to see how they do it but I guess this might be *fine* for now

Reply:
Yeah i let them know about the issue for mounting back in january but they dont respond so shrug